doc/secure_boot/hld_secure_boot.md

```diff
@@ -0,0 +1,345 @@
+
+SONiC is using the initrd, so we will remain it as in the original repo unsigned. (In order to sign it is required to use a gpg key, but there is not really a use case for that. (Probably future work.)
+
+Grub.cfg - The grub.cfg file is the GRUB configuration file. It is generated by the grub2-mkconfig program using a set of primary configuration files and the grub default file as a source for user configuration specifications. The /boot/grub2/grub.cfg file is first generated during Linux installation and regenerated when a new kernel is installed.
+
+The grub.cfg file contains Bash-like code and a list of installed kernels in an array ordered by sequence of installation. For example, if you have four installed kernels, the most recent kernel will be at index 0, the previous kernel will be at index 1, and the oldest kernel will be at index 3.
+
+(Not modification required)
+
+##### First instance of trust
+SB chain of trust has different options for this first instance of trust.
+This HLD show 2 options:
+1.  BIOS/UEFI as the first instance of trust, vendors can store their public keys in the UEFI DB, by doing that, they enforce to boot images that were only signed with a private key that matches the vendor's public key.
+the diagram can be founded in the "Run-time verification flow diagram with BIOS as the first instance of trust".
+This is a generic way that Linux support using the secure boot feature and we add support to it as well.
+Linux Debian SB link: https://wiki.debian.org/SecureBoot
+
+2. HW as a root of trust, full HW root of trust is reliant on the HW vendor and production flow since customers can add this by themselves depending on the HW platform they have.
+So in this HLD, we showed the diagram "Run-time verification flow diagram with HW root of trust", but, this implementation should be decided by the vendor.
+
+####  1.7.3. <a name='SignFlowdiagramdescription:'></a>Sign Flow diagram description:
+Sign flow occurs when building the SONiC image.
+In the Build process, we need to sign the components shown in the Design Diagram chapter (shim, grub, vmlinuz, kernel modules) and do some modifications in the Linux kernel, that in order to pass the runtime secure boot verification.
+
+This chapter will be described which SONiC modification is required.
+
+Compilation Flag added:
+- SB_DEV_KEYS_PATH – have the path where dev private key stored
+- SB_BUILD – have 3 options
+  - No sign
+  - Dev (default)
+  - Production
+- SB_KERNEL_MOD_PUB_KEY – the path of the public key that will be embedded in VMLINUZ for verifying KO.
+These 3 flags will be saved in this file: SONiC/rules/config
+
+##### Sign keys
+The sign algorithm is RSA, is an asymmetric algorithm, so the sign flow is signing with private keys, and the verification flow is verifying with the public keys.
+The are 2 kinds of keys, development key and production keys.
+
+The production keys will be stored according to the design of the signing server
```

Comment:
What kind of signing server is it? is it kind of Vault server? How does SONiC communicate with signing server? Is it an API based

Reply:
any vendor can create a communitcaction from SONiC to the production server in the way that he like.
the Vendor should write his communication in this script name: signing_secure_boot_production.py 
(explained in chapter 1.6. Architecture Design)
after selecting the make flag SB_BUILD production